corehq/blobs/migrate.py

```diff
@@ -259,45 +275,81 @@ def encode_content(data):
             data = data.encode("utf-8")
         return b64encode(data)
 
-    print("Loading documents: {}...".format(", ".join(doc_type_map)))
-    total = 0
+    total = sum(get_doc_count_by_type(couchdb, doc_type)
+                for doc_type in doc_type_map)
+    print("Migrating {} documents: {}...".format(
+        total,
+        ", ".join(sorted(doc_type_map))
+    ))
+    migrated = 0
+    skipped = 0
+    visited = 0
     start = datetime.now()
-    load_attachments = getattr(migrate_func, "load_attachments", False)
-    with open(filename, 'w') as f:
-        for doc in get_all_docs_with_doc_types(couchdb, list(doc_type_map)):
-            if doc.get(migrate_func.blobs_key):
+    doc_migrator = doc_migrator_class()
+    load_attachments = doc_migrator.load_attachments
+    blobs_key = doc_migrator.blobs_key
+    iter_key = slug + "-blob-migration"
+    docs_by_type = ResumableDocsByTypeIterator(couchdb, doc_type_map, iter_key)
+    if reset:
+        docs_by_type.discard_state()
+
+    with open(filename, 'wb') as f:
+        for doc in docs_by_type:
+            visited += 1
+            if doc.get(blobs_key):
                 if load_attachments:
-                    obj = doc_type_map[doc["doc_type"]].wrap(doc)
-                    fetch_attachment = super(BlobMixin, obj).fetch_attachment
+                    obj = BlobHelper(doc, couchdb)
                     doc["_attachments"] = {
                         name: {
                             "content_type": meta["content_type"],
-                            "content": encode_content(fetch_attachment(name)),
+                            "content": obj.fetch_attachment(name),
                         }
                         for name, meta in doc["_attachments"].items()
                     }
-                f.write('{}\n'.format(json.dumps(doc)))
-                total += 1
-                if total % 100 == 0:
-                    print("Loaded {} documents in {}"
-                          .format(total, datetime.now() - start))
+                    # make copy with encoded attachments for JSON dump
+                    backup_doc = dict(doc)
+                    backup_doc["_attachments"] = {
+                        name: {
+                            "content_type": meta["content_type"],
+                            "content": encode_content(meta["content"]),
+                        }
+                        for name, meta in doc["_attachments"].items()
+                    }
+                else:
+                    backup_doc = doc
+                f.write('{}\n'.format(json.dumps(backup_doc)))
+                f.flush()
+                ok = doc_migrator.migrate(doc, couchdb)
```

Comment:
Documents are now migrated as they are loaded from the database rather than loading all documents into a file and then migrating the file in a separate phase. This has several advantages:

- The migration can be stopped and resumed without incurring a massive reload of unmigrated documents.
- If the migration log gets too big, the migration can be stopped to allow space to be freed up before resuming the migration.
- Less chance of concurrent modification errors because it is a relatively short time from when a document is loaded from couch to when it is migrated and finally saved.

As before, documents, including all attachments, are serialized to a migration log before they are modified as a safety precaution in case something goes wrong with the migration.